03-wordemb-pytorch/wordemb-skip.py

```diff
@@ -0,0 +1,122 @@
+import math
+import time
+import random
+import torch
+import torch.nn.functional as F
+
+
+class Skip(torch.nn.Module):
+    def __init__(self, nwords, emb_size):
+        super(Skip, self).__init__()
+
+        """ word embeddings """
+        self.word_embedding = torch.nn.Embedding(nwords, emb_size)
+        """ context embeddings"""
+        self.context_embedding = torch.nn.Embedding(nwords, emb_size)
+
+    def forward(self, word_pos, context_pos):
+        embed_word = self.word_embedding(word_pos)    # 1 * emb_size
+        embed_context = self.context_embedding(context_pos)  # 1 * emb_size
+        score = torch.mul(embed_word, embed_context)
+        score = torch.sum(score, dim=1)
+        log_target = -1 * F.logsigmoid(score).squeeze()
+        return log_target
+
+
+N = 2  # length of window on each side (so N=2 gives a total window size of 5, as in t-2 t-1 t t+1 t+2)
+EMB_SIZE = 128  # The size of the embedding
+
+embeddings_location = "embeddings.txt"  # the file to write the word embeddings to
+labels_location = "labels.txt"  # the file to write the labels to
+
+# We reuse the data reading from the language modeling class
+w2i = defaultdict(lambda: len(w2i))
+S = w2i["<s>"]
+UNK = w2i["<unk>"]
+
+
+def read_dataset(filename):
+    with open(filename, "r") as f:
+        for line in f:
+            yield [w2i[x] for x in line.strip().split(" ")]
+
+
+# Read in the data
+train = list(read_dataset("../data/ptb/train.txt"))
+w2i = defaultdict(lambda: UNK, w2i)
+dev = list(read_dataset("../data/ptb/valid.txt"))
+i2w = {v: k for k, v in w2i.items()}
+nwords = len(w2i)
+
+with open(labels_location, 'w') as labels_file:
+    for i in range(nwords):
+        labels_file.write(i2w[i] + '\n')
+
+# initialize the model
+model = Skip(nwords, EMB_SIZE)
+optimizer = torch.optim.SGD(model.parameters(), lr=0.1)
+
+type = torch.LongTensor
+use_cuda = torch.cuda.is_available()
+
+if use_cuda:
+    type = torch.cuda.LongTensor
+    model.cuda()
+
+
+# Calculate the loss value for the entire sentence
+def calc_sent_loss(sent, inference=False):
+    # add padding to the sentence equal to the size of the window
+    # as we need to predict the eos as well, the future window at that point is N past it
+
+    # Step through the sentence
+    total_loss = 0
+    for i, word in enumerate(sent):
+        c = torch.tensor([word]).type(type)     # This is tensor for center word
+        for j in range(1, N + 1):
+            for direction in [-1, 1]:
+                context_id = sent[i + direction * j] if 0 <= i + direction * j < len(sent) else S
+                context = torch.tensor([context_id]).type(type)   # Tensor for context word
+                loss = model(c, context)
+                if not inference:
+                    # Back prop while training only
+                    optimizer.zero_grad()
+                    loss.backward()
+                    optimizer.step()
+                total_loss += loss.data.cpu().item()
```

Comment:
A nit pick: but for single valued tensors, you can simply do `loss.item()`, you don't need to transfer the data to cpu.

Reply:
Changed to .item()